### PR TITLE
 Understanding shared memory programming with Pthreads and OpenMP 

### LINKED ISSUE #6135: Understanding shared memory programming with Pthreads and OpenMP
#### Proposed title of the article
Understanding shared memory programming with Pthreads and OpenMP.

#### Proposed article introduction
Data is sent between applications more efficiently via shared memory. Depending on the circumstances, programs may execute on a single processor or numerous processors. Shared memory refers to communication inside a single program, such as between its innumerable threads.

There may be many threads in the process. The lightweight process is another term for a thread. By splitting a process into many threads, parallelism may be achieved. Multiple tabs in a browser, for example, may each be a separate thread. MS Word uses numerous threads to format text, handle inputs, and so on. Below are some additional benefits of multithreading.

This tutorial will use the Pthreads and OpenMP programming paradigms to demonstrate how to create parallel applications.

#### Key takeaways 
By the end of this article, the reader will have understood:

- the system architecture.
- what shared memory, threads and processes are. 
- the parallel computing: shared memory models.
- what Pthread is and why use Pthread.
- what OpenMP is and why use OpenMP.
- how threads work, their safety and limits
- the Pthreads API 
- working with OpenMP and Pthreads.

#### Article quality
This article aims to go further into this topic while also presenting its principles in a straightforward, understandable manner for newcomers.

References

#### References
N/A

### PR DESCRIPTION
**Pull request by carteblanchemunene**
#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

#### Topic suggestion this closes

This closes #6135

### CODE REVIEW

**Comment by Dawe-7** [comment]
@jerimkaura, i didn't know this is a duplicate PR you were reviewing. Kindly continue with the review.

**Comment by jerimkaura** [comment]
@Dawe-7 

You can just proceed with the review. I'm kind of held up.

**Comment by Dawe-7** [comment]
> @Dawe-7
> 
> You can just proceed with the review. I'm kind of held up.

Okay  proceeding with the review.

**Review by Dawe-7 — COMMENTED** [review]
@carteblanchemunene, kindly check the file path (URL) of your article in VS Code. It could bring about merge conflicts.

File: content/articles/understanding-shared-memory-programming-with-pthreads-and-openmp/index.md [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+3. `Building a graphical user interface` - Graphical user interface apps expect a request to do a window portion. If it is pre-occupied, the window will be blank. In this scenario, having one thread handle the windowing system messages and requests is prudent (as well as user input). If an operation takes more than 0.2 seconds, it is sent to another thread.
+
+### OpenMP and its use
+OpenMP is an SMP programming package. OpenMP threads share memory and data. This includes C++ and FORTRAN. The OpenMP header file is omp.h.
+
+An OpenMP application's parts might be sequential or parallel. For example, an OpenMP program often begins with a sequential selection that sets up the environment and initializes the variables.
+
+When an OpenMP application is launched, it will utilize one thread (in the sequential portions) and numerous threads (in the parallel sections).
+
+The primary thread is the one that goes all the way from the beginning to the conclusion. On the other hand, the secondary threads will fork due to the parallel parts of the program.
```
Comment (Dawe-7) [inline content/articles/understanding-shared-memory-programming-with-pthreads-and-openmp/index.md:244]:
```suggestion
The primary thread is the thread that travels from the start to the end. On the other hand, the secondary threads will fork due to the parallel parts of the program.
```

File: content/articles/understanding-shared-memory-programming-with-pthreads-and-openmp/index.md [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+
+The primary thread is the one that goes all the way from the beginning to the conclusion. On the other hand, the secondary threads will fork due to the parallel parts of the program. 
+
+A particular directive marks a block of code executed in a parallel program. This directive will induce secondary threads to arise when the execution reaches a parallel portion (indicated by omp pragma). 
+
+Each thread runs a similar piece of code. It joins the master when finished. When all threads have been terminated, the master resumes programming in the parallel part.
+
+The runtime library's `omp get thread num()` method finds each thread's ID. The main thread ID is 0.
+
+Use OpenMP directives to:
```
Comment (Dawe-7) [inline content/articles/understanding-shared-memory-programming-with-pthreads-and-openmp/index.md:252]:
```suggestion
We use OpenMP directives to:
```

File: content/articles/understanding-shared-memory-programming-with-pthreads-and-openmp/index.md [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+- Define a parallel zone.
+- Define whether parallel section variables are private or shared.
+- Define synchronization of threads.
+- Define loop parallelization.
+- Indicate the work's threading.
+   
+#### Creating a thread using OpenMP
+Here is what you should do to create a thread using OpenMP:
```
Comment (Dawe-7) [inline content/articles/understanding-shared-memory-programming-with-pthreads-and-openmp/index.md:260]:
```suggestion
Here is the code to create a thread using OpenMP:
```

File: content/articles/understanding-shared-memory-programming-with-pthreads-and-openmp/index.md [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+The output is:
+
+```bash
+Hi and welcome to Section Engineering education.
+
+Hi and welcome to Section Engineering education.
+```
+
+#### Parallelizing loops
+OpenMP takes care of parallelizing loops with only a few parameters and a loop definition.
```
Comment (Dawe-7) [inline content/articles/understanding-shared-memory-programming-with-pthreads-and-openmp/index.md:298]:
```suggestion
OpenMP takes care of parallelizing loops with only a few parameters and a loop definition. The directive, work-sharing concept, must be in a parallel section.
```

File: content/articles/understanding-shared-memory-programming-with-pthreads-and-openmp/index.md [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+```bash
+Hi and welcome to Section Engineering education.
+
+Hi and welcome to Section Engineering education.
+```
+
+#### Parallelizing loops
+OpenMP takes care of parallelizing loops with only a few parameters and a loop definition. 
+
+The directive, work-sharing concept, must be in a parallel section:
```
Comment (Dawe-7) [inline content/articles/understanding-shared-memory-programming-with-pthreads-and-openmp/index.md:300]:
```suggestion
```

File: content/articles/understanding-shared-memory-programming-with-pthreads-and-openmp/index.md [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+
+The `#pragma omp` command distributes the loop across threads. It must be used with another block of code:
+
+
+Let us look at an example program that adds all elements in an array. Save the code as `third.c`.
+
+```C++
```
Comment (Dawe-7) [inline content/articles/understanding-shared-memory-programming-with-pthreads-and-openmp/index.md:307]: Check on your code indentation

**Review by Dawe-7 — CHANGES_REQUESTED** [review]
Hello @carteblanchemunene, your article is still coming in with a handful of errors and it's a duplicate PR. Always make sure it is fully polished when you create a new PR to avoid several revisions. Kindly implement the suggestions below. 

Some images are missing as well, make sure to add them to your article. Ensure you have run a grammar check and clarify any unclear sentences (check on readability) in the article.

**Comment by Dawe-7** [comment]
@carteblanchemunene, kindly ensure all the changes have been effected and the article is ready for final review. We will be closing this PR until to clear up the queue.

Ping me after you've created a duplicate PR.

**Pull request closed without merging**